src/jitlayers.cpp

```diff
@@ -2315,54 +2007,225 @@ void JuliaOJIT::enablePerfJITEventListener()
+    auto AddSyms = [&](auto Symbols) {
+        for (auto S : Symbols)
+            if (S->getName())
+                Syms[S->getName()] = S;
+    };
+    AddSyms(G.defined_symbols());
+    AddSyms(G.external_symbols());
+    return Syms;
 }
+
+void JuliaOJIT::linkOutput(orc::MaterializationResponsibility &MR, MemoryBufferRef ObjBuf,
+                           jitlink::LinkGraph &G, std::unique_ptr<jl_linker_info_t> Info)
+{
+    auto Syms = linkGraphSymbols(G);
+
+    // Rename the defined CI functions.
+    std::lock_guard Lock{LinkerMutex};
+    auto RenameDef = [&](const SymbolStringPtr &Orig, const SymbolStringPtr &Dest) {
+        Syms.at(Orig)->setName(Dest);
+    };
+    for (auto &[CI, Funcs] : Info->ci_funcs) {
+        auto &S = CISymbols.at(CI);
+        if (Funcs.invoke)
+            RenameDef(Funcs.invoke, S.invoke);
+        if (Funcs.specptr)
+            RenameDef(Funcs.specptr, S.specptr);
+    }
+
+    // Rename referenced CIs in the workqueue.
+    for (auto &[Call, T] : Info->call_targets) {
+        auto [CI, API] = Call;
+        if (!Syms.contains(T))
+            continue;
+        JL_GC_PROMISE_ROOTED(CI);
+        Syms.at(T)->setName(linkCallTarget(CI, API));
+    }
+
+    // Rename globals and add mappings
+    // TODO: don't leak when we have a way to GC code
+#ifdef __clang_analyzer__
+    [[clang::suppress]]
 #endif
+    void **Ptrs = new void *[Info->global_targets.size()];
+    size_t i = 0;
+    orc::SymbolMap GlobalSyms;
+    for (auto &[Addr, Orig] : Info->global_targets) {
+        auto Sym = ES.intern(Names(*Orig, "#"));
+        auto It = Syms.find(Orig);
+        if (It == Syms.end())
+            continue;
+        It->second->setName(Sym);
+        Ptrs[i] = Addr;
+        GlobalSyms[Sym] = {ExecutorAddr::fromPtr(Ptrs + i), JITSymbolFlags::Exported};
+        ++i;
+        ++LinkedGlobals;
+    }
+    cantFail(JD.define(orc::absoluteSymbols(std::move(GlobalSyms))));
 
-const DataLayout& JuliaOJIT::getDataLayout() const
+    DebuginfoPlugin->notifyMaterializingWithInfo(MR, G, ObjBuf, std::move(Info));
+}
+
+// Must hold LinkerMutex.
+orc::SymbolStringPtr JuliaOJIT::linkCallTarget(jl_code_instance_t *CI, jl_invoke_api_t API)
 {
-    return DL;
+    auto It = CISymbols.find(CI);
+    if (It != CISymbols.end())
+        return It->second.specptr;
+
+    CISymbolPtr *Sym = linkCISymbol(CI);
+
+    orc::SymbolStringPtr Result;
+    CISymbolPtr Trampoline;
+    std::unique_ptr<jl_codegen_output_t> Out;
+
+    if (!Sym) {
+        // The target CI was not compiled when we observed it, so generate a
+        // tojlinvoke trampoline that will cause it to be compiled.
+        // TODO: replace this with a GOT/PLT mechanism that avoids the jl_invoke
+        // after it has been compiled.
+        Out = std::make_unique<jl_codegen_output_t>("tojlinvoke", getDataLayout(),
+                                                    getTargetTriple());
+        Function *F = emit_tojlinvoke(CI, StringRef(), *Out);
```

Comment:
We should review to make sure this cannot acquire any mutexes or safepoints, since that would cause a lock inversion violation with the LinkerMutex

(e.g. annotate the containing function JL_NOTSAFEPOINT to indicate that the caller is permitted to hold a lock)